### ISSUE TITLE
Detailing HTTP-Signature based authentication for Solid

### ISSUE THREAD

**Issue by bblfish** [solid/authentication-panel#18]
I think it would be useful to detail [HTTP-Signatures](https://w3c-dvcg.github.io/http-signatures/) based authentication for Solid. I [wrote an implementation for it](https://w3c-dvcg.github.io/http-signatures/) a few years ago for the server, and one before that for the client using Web-Crypto.

Not much is needed in addition to the existing spec. I can think of the following:
 1. A definititon of a Web-Key: a URL that names a public key, that can be either
   a) located in the WebID doc
   b) located elsewhere in case there is a desire to not tie a key to a webid.
2. A verification procedure where the server fetches the public WebKey to check the client
   a) if the wACL identifies that Agent via that WebKey that that is all that is needed
   b) if the wACL identifies the Agent via a WebID then the client needs to send the WebId in the request. If the situation is 1a above, then that is all that is needed. If 1b above then there needs to be a proof tying the WebId to the WebKey.

**Comment by bblfish** [comment]
This protocol was discussed with @elf-pavlik and @zenomt on Wednesday 11 Sept 2019 and [notes left on hackmd.io](https://hackmd.io/bvw37jzHSRiA35TWrKBF3A). Will try to distill some of the thoughts here.

**Comment by bblfish** [comment]
There were a couple of discussions on the Solid group of the intiial proposal, with meeting notes [listed here](https://github.com/solid/authentication-panel/issues/18#issuecomment-530420480).
The first was with a longer but with only 3 members present, the second was shorter but with a number of people of which the CTO of Inrupt.
Perhaps it would be worth chatting to a some people here about those meetings, so I can understand the larger context in which "Signing HTTP Messages" should be understood.

**Comment by dmitrizagidulin** [comment]
One of the things that's worth noting, about any sort fo HTTP-Sig based proposals, is that it's currently used almost exclusively for secure clients (meaning, traditional server-side applications) and for server-to-server authentication. By itself, it is not appropriate for authentication of public clients (SPA / in-browser Javascript apps).

**Comment by bblfish** [comment]
Here is a way to resolve the problem.

1. We can give the user pod `me.me` host an authentication JavaScript App. This App will be in charge of keeping keys using [JS Crypto's generateKey](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/generateKey) securely in the end-user's browser, in a way that not even the App can access it. 
2. the WebID Profile can link to this authorization app perhaps with something like
```Turtle
   <#me> :authenticationApp </safe/authentication.html> . //which contains the JS.
```
3. An App from `app.net` could then on finding out the WebID of the user find the authenticationApp page, open that in a new frame.
4. Using [Window.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage), `app.net` can communicate with the Authentication Window and request HTTP headers to be signed.  

One could also go the other way around, and register apps in one's authentication app, and have that open a browser window in which the app occurs.

**Comment by bblfish** [comment]
Regarding the question as to what the problem with HTTP-Signature is felt to be @dmitrizagidulin answered on gitter that it had to do with the [problem of the Confused Deputy](https://gitter.im/solid/app-authorization-panel?at=5d8b8ea1a8c2c26e9f9bad10), which is what is behind the emergence of Capability Systems, and that these arise when 3 or more agents are involved in a secure communication. Here we have the App as the extra agent. This may be resolvable by being able to identify the app.

I'll look into the confused deputy and capabilities vs ACL debate to which Dmitry provided ample pointers.

**Comment by elf-pavlik** [comment]
@bblfish given we have merged #20 I think we can close this issue and follow up with focused issues which can get addressed by focused PRs.
